Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -0,0 +1,386 @@
+              "finishes");
+  HARD_ASSERT(state_ == State::Open, "Read called when the stream is not open");
+
+  has_pending_read_ = true;
+  Execute<StreamRead>();
+}
+
+void GrpcStream::Write(grpc::ByteBuffer&& message) {
+  HARD_ASSERT(state_ == State::Open,
+              "Write called when the stream is not open");
+  BufferedWrite(std::move(message));
+}
+
+void GrpcStream::Finish() {
+  if (state_ == State::NotStarted) {
+    return;
+  }
+
+  HARD_ASSERT(state_ < State::Finishing, "Finish called twice");
+  state_ = State::Finishing;
+
+  buffered_writer_.reset();
+  // Important: since the stream always has a pending read operation,
+  // cancellation has to be called, or else the read would hang forever, and
+  // finish operation will never get completed (an operation cannot be completed
+  // before all previously-enqueued operations complete).
+  //
+  // On the other hand, when an operation fails, cancellation should not be
+  // called, otherwise the real failure cause will be overwritten by status
+  // "canceled".
+  Execute<ClientInitiatedFinish>();
+}
+
+void GrpcStream::WriteAndFinish(grpc::ByteBuffer&& message) {
+  if (state_ < State::Open) {
+    // Ignore the write part if the call didn't have a chance to open yet.
+    Finish();
+    return;
+  }
+
+  HARD_ASSERT(buffered_writer_,
+              "Write requested when there is no valid buffered_writer_");
+  state_ = State::LastWrite;
+  // Write the last message as soon as possible by discarding anything else that
+  // might be buffered.
+  buffered_writer_->DiscardUnstartedWrites();
+  BufferedWrite(std::move(message));
+}
+
+void GrpcStream::BufferedWrite(grpc::ByteBuffer&& message) {
+  HARD_ASSERT(buffered_writer_,
+              "Write requested when there is no valid buffered_writer_");
+  StreamWrite* write_operation = MakeOperation<StreamWrite>(std::move(message));
+  buffered_writer_->EnqueueWrite(write_operation);
+}
+
+bool GrpcStream::SameGeneration() const {
+  return generation_ == observer_->generation();
+}
+
+// Callbacks
+
+void GrpcStream::OnStart() {
+  state_ = State::Open;
+  buffered_writer_.emplace(BufferedWriter{});
+
+  if (SameGeneration()) {
+    observer_->OnStreamStart();
+    // Start listening for new messages.
+    Read();
+  }
+}
+
+void GrpcStream::OnRead(const grpc::ByteBuffer& message) {
+  has_pending_read_ = false;
+
+  if (SameGeneration()) {
+    observer_->OnStreamRead(message);
+    if (state_ == State::Open) {
+      // Continue waiting for new messages indefinitely as long as there is an
+      // interested observer and the stream is open.
+      Read();
+    }
+  }
+}
+
+void GrpcStream::OnWrite() {
+  if (state_ == State::LastWrite && buffered_writer_->empty()) {
+    // Final write succeeded.
+    Finish();
+    return;
+  }
+  if (state_ <= State::LastWrite) {
+    buffered_writer_->DequeueNextWrite();
+  }
+
+  if (SameGeneration()) {
+    observer_->OnStreamWrite();
```

Comment:
There actually is no logic in `WatchStream`/`WriteStream` that depends on this callback. I originally added it for symmetry, but now that I think about it, perhaps it would be better not to expose this to the observer at all. It appears that this method largely exist due to GRPC not allowing more than one pending write; since `GrpcStream` implements buffering and hides this inconvenience from the public API, perhaps this callback should be hidden as well.